### ISSUE TITLE
distsql: support ARRAY_AGG in distsql

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#15939]
We currently don't support arrays in DistSQL streams. The main thing that's missing is a way to encode them; this will have to be implemented to support arrays in table columns (see #15818). Once that's done we should enable support in DistSQL as well.

CC @asubiotto

**Comment by RaduBerinde** [comment]
Part of this work we should also implement `ARRAY_AGG` in distsql.

**Comment by cuongdo** [comment]
Given that we currently only support arrays for virtual schemas, is this currently affecting anything? Or is this a natural part of our work to support ARRAYs?

**Comment by RaduBerinde** [comment]
We support arrays in select expressions (and there is `ARRAY_AGG`).

**Comment by RaduBerinde** [comment]
Hi @a6802739 - to work on this we need array encoding code to be in place. This is only in the RFC stage at this point (#16172).

**Comment by a6802739** [comment]
@RaduBerinde  Thank you, I will wait until it's ready.

**Comment by cuongdo** [comment]
@RaduBerinde can you reassign?

**Comment by RaduBerinde** [comment]
@asubiotto I'll put this on your plate for now. We may need to move to 1.2 (not sure what is the current state of array encoding for table columns).

**Comment by jordanlewis** [comment]
cc @justinj

**Comment by arjunravinarayan** [comment]
@cuongdo This is not going to make it into 1.1, @asubiotto's other big priority is getting external storage for HashJoiners and there isn't time for both big projects to make it in in 1.1, so lets move it to 1.2 unless we can get more hands on deck between now and the 1.1 feature freeze.

**Comment by RaduBerinde** [comment]
Fixed in #18243.

**Comment by RaduBerinde** [comment]
> Part of this work we should also implement ARRAY_AGG in distsql.

This is still to be done.

**Comment by vivekmenezes** [comment]
assigning this issue to abhishek. Can't assign at the moment because he's still not a contributor

**Comment by abhimadan** [comment]
I think that adding `ARRAY_AGG` support should only involve removing the explicit check for it in the physical planner, and adding it to the `AggregatorSpec_Func` proto enum. However, when I tried to do this, I ran into the CLI error `pq: unsupported result type: anyelement`.

After some investigating, I found that the aggregator processor calls `tree.Builtin.FixedType()`, which in turn calls `returnTypeToFixedType()`. This function relies on the fact that its `ReturnTyper` argument returns a valid type when passed an empty expression list. However, `ARRAY_AGG` does not follow this assumption ([see here](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/aggregate_builtins.go#L78-L83)). We can actually see this bug manifest in the user docs ([see here](https://www.cockroachlabs.com/docs/stable/functions-and-operators.html#aggregate-functions)), where `ARRAY_AGG` returns `anyelement` instead of an array.

I'm getting a PR ready to fix this issue.
